src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1718,16 +1873,70 @@ WorkerAggregateWalker(Node *node, List **newExpressionList)
 /*
  * WorkerAggregateExpressionList takes in the original aggregate function, and
  * determines the transformed aggregate functions to execute on worker nodes.
- * The function then returns these aggregates in a list.
+ * The function then returns these aggregates in a list. It also creates
+ * group by clauses for newly added targets to be placed in the extended operator
+ * node.
  */
 static List *
-WorkerAggregateExpressionList(Aggref *originalAggregate)
+WorkerAggregateExpressionList(Aggref *originalAggregate,
+							  WorkerAggregateWalkerContext *walkerContext)
 {
 	AggregateType aggregateType = GetAggregateType(originalAggregate->aggfnoid);
 	List *workerAggregateList = NIL;
 
 	if (aggregateType == AGGREGATE_COUNT && originalAggregate->aggdistinct &&
-		CountDistinctErrorRate != DISABLE_DISTINCT_APPROXIMATION)
+		CountDistinctErrorRate == DISABLE_DISTINCT_APPROXIMATION &&
+		walkerContext->repartitionSubquery)
+	{
+		Aggref *aggregate = (Aggref *) copyObject(originalAggregate);
+		List *aggTargetEntryList = aggregate->args;
+		TargetEntry *distinctTargetEntry = (TargetEntry *) linitial(aggTargetEntryList);
+		List *columnList = pull_var_clause_default((Node *) distinctTargetEntry);
+		ListCell *columnCell = NULL;
+		List *groupRefList = NIL;
+
+		foreach(columnCell, columnList)
+		{
+			TargetEntry *newTargetEntry = makeNode(TargetEntry);
+			Var *column = (Var *) lfirst(columnCell);
+			SortGroupClause *newDistinctGroupBy = NULL;
+			Oid opLt = InvalidOid;
+			Oid opEq = InvalidOid;
+			Oid opGt = InvalidOid;
+			bool hashable = false;
+
+			newTargetEntry->expr = copyObject(column);
+			newTargetEntry->resno = 0;
+			newTargetEntry->resname = NULL;
+			newTargetEntry->ressortgroupref = walkerContext->nextSortGroupRefIndex;
```

Comment:
it also depends on the name of the original target entry in the query, this name is not available in the inner function. we might send it in inside walker context, but it would introduce more complexity than it resolves.